### ISSUE TITLE
Possible to estimate gradient numerically?

### ISSUE THREAD

**Issue by jsimons8** [yixuan/LBFGSpp#4]
Hi Xuan, this is an amazing library; it is especially helpful to have a functor construct as most objective functions do have so many other parameters that are not the argument.

Just a quick question, is it ever on the cards to augment this library with numerically estimated gradients? I know this issue had been closed already as the original person interested in this found analytical derivatives but I really do have a problem that doesn't have analytical derivatives (only to first order and they're rubbish). Thanks so much!

**Comment by DirkToewe** [comment]
@jsimons8 You could build a Functor that takes another Functor `f` and computes the gradients for is using [Finite Differences](https://en.wikipedia.org/wiki/Finite_difference).
```C++
template<typename Functor>
class NumDiff
{
private:
  const Functor f;
  const double rtol, atol;
public:
  NumDiff(
    Functor f,
    double rtol=std::sqrt( std::numeric_limits<double>::epsilon() ),
    double atol=std::sqrt( std::numeric_limits<double>::epsilon() )
  ): f(f), rtol(rtol), atol(atol) {}

  double operator()(const VectorXd& x, VectorXd& grad)
  {
    VectorXd dx = x;
    const double fx = f(x);

    for( ptrdiff_t i=x.rows(); i-- > 0; )
    {
      const double h = std::abs(x(i))*rtol + atol;
      dx(i) = x(i) + h;
      grad(i) = ( f(dx) - fx ) / h;
      dx(i) = x(i);
    }

    return fx;
  }
};
```
There is a multitude of Finite Difference methods. If You need a better approximation, You can try using [higher order Central Differences](http://mathfaculty.fullerton.edu/mathews/n2003/differentiation/NumericalDiffProof.pdf). But Finite Differences might not work out as well as You might hope for, for two reasons:
  * If Your function doesn't have analytical derivatives, it might not be continuously differentiable and therefore not suited for an (L)BFGS optimizer.
  * If Your function `f` takes an `n` dimensional input, Finite Differences needs to call `f` `n` times every time the gradients are estimated. In my experiences that makes optimization unbearably slow.

As @yixuan suggested in [issue#1](https://github.com/yixuan/LBFGSpp/issues/1), You could try derivative-free methods like [Nelder-Mead](https://en.wikipedia.org/wiki/Nelder%E2%80%93Mead_method).

You can also try to use one of the many Machine Learning frameworks like [Tensorflow](https://www.tensorflow.org/) or [PyTorch](https://pytorch.org/) to compute the gradients automatically via backpropagation.
